app/models/article.rb

```diff
@@ -435,6 +435,19 @@ def plain_html
     doc.to_html
   end
 
+  def followers
+    # This will return an array, but the items will NOT be ActiveRecord objects.
+    # The followers may also occasionally be nil because orphaned follows can possibly exist in the database.
+    followers = user.followers_scoped.where(subscription_status: "all_articles").map(&:follower)
+
+    if organization_id
+      org_followers = organization.followers_scoped.where(subscription_status: "all_articles")
+      followers += org_followers.map(&:follower)
+    end
+
+    followers.uniq.compact
+  end
+
```

Comment:
I'm proposing an optimization on this method. 

Currently, after loading an article belonging to an organization in memory in the console, it runs up to 6 queries:

```sql
[10] pry(main)> art.followers
  User Load (14.4ms)  SELECT "users"."id", "users"."apple_created_at", "users"."apple_username", "users"."articles_count", "users"."badge_achievements_count", "users"."blocked_by_count", "users"."blocking_others_count", "users"."checked_code_of_conduct", "users"."checked_terms_and_conditions", "users"."comments_count", "users"."config_font", "users"."config_navbar", "users"."config_theme", "users"."confirmation_sent_at", "users"."confirmation_token", "users"."confirmed_at", "users"."created_at", "users"."credits_count", "users"."current_sign_in_at", "users"."current_sign_in_ip", "users"."display_announcements", "users"."display_sponsors", "users"."editor_version", "users"."email", "users"."email_badge_notifications", "users"."email_comment_notifications", "users"."email_community_mod_newsletter", "users"."email_connect_messages", "users"."email_digest_periodic", "users"."email_follower_notifications", "users"."email_membership_newsletter", "users"."email_mention_notifications", "users"."email_newsletter", "users"."email_tag_mod_newsletter", "users"."email_unread_notifications", "users"."encrypted_password", "users"."experience_level", "users"."export_requested", "users"."exported_at", "users"."facebook_created_at", "users"."facebook_username", "users"."failed_attempts", "users"."feed_fetched_at", "users"."feed_mark_canonical", "users"."feed_referential_link", "users"."feed_url", "users"."following_orgs_count", "users"."following_tags_count", "users"."following_users_count", "users"."github_created_at", "users"."github_repos_updated_at", "users"."github_username", "users"."inbox_guidelines", "users"."inbox_type", "users"."invitation_accepted_at", "users"."invitation_created_at", "users"."invitation_limit", "users"."invitation_sent_at", "users"."invitation_token", "users"."invitations_count", "users"."invited_by_id", "users"."invited_by_type", "users"."last_article_at", "users"."last_comment_at", "users"."last_followed_at", "users"."last_moderation_notification", "users"."last_notification_activity", "users"."last_onboarding_page", "users"."last_reacted_at", "users"."last_sign_in_at", "users"."last_sign_in_ip", "users"."latest_article_updated_at", "users"."locked_at", "users"."mobile_comment_notifications", "users"."mod_roundrobin_notifications", "users"."monthly_dues", "users"."name", "users"."old_old_username", "users"."old_username", "users"."onboarding_package_requested", "users"."organization_info_updated_at", "users"."payment_pointer", "users"."permit_adjacent_sponsors", "users"."profile_image", "users"."profile_updated_at", "users"."rating_votes_count", "users"."reaction_notifications", "users"."reactions_count", "users"."registered", "users"."registered_at", "users"."remember_created_at", "users"."remember_token", "users"."reputation_modifier", "users"."reset_password_sent_at", "users"."reset_password_token", "users"."saw_onboarding", "users"."score", "users"."secret", "users"."sign_in_count", "users"."signup_cta_variant", "users"."spent_credits_count", "users"."stripe_id_code", "users"."subscribed_to_user_subscriptions_count", "users"."twitter_created_at", "users"."twitter_username", "users"."unconfirmed_email", "users"."unlock_token", "users"."unspent_credits_count", "users"."updated_at", "users"."username", "users"."welcome_notifications", "users"."workshop_expiration" FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 4002], ["LIMIT", 1]]
  Follow Load (0.6ms)  SELECT "follows".* FROM "follows" WHERE "follows"."followable_id" = $1 AND "follows"."followable_type" = $2 AND "follows"."subscription_status" = $3  [["followable_id", 4002], ["followable_type", "User"], ["subscription_status", "all_articles"]]
  User Load (0.6ms)  SELECT "users"."id", "users"."apple_created_at", "users"."apple_username", "users"."articles_count", "users"."badge_achievements_count", "users"."blocked_by_count", "users"."blocking_others_count", "users"."checked_code_of_conduct", "users"."checked_terms_and_conditions", "users"."comments_count", "users"."config_font", "users"."config_navbar", "users"."config_theme", "users"."confirmation_sent_at", "users"."confirmation_token", "users"."confirmed_at", "users"."created_at", "users"."credits_count", "users"."current_sign_in_at", "users"."current_sign_in_ip", "users"."display_announcements", "users"."display_sponsors", "users"."editor_version", "users"."email", "users"."email_badge_notifications", "users"."email_comment_notifications", "users"."email_community_mod_newsletter", "users"."email_connect_messages", "users"."email_digest_periodic", "users"."email_follower_notifications", "users"."email_membership_newsletter", "users"."email_mention_notifications", "users"."email_newsletter", "users"."email_tag_mod_newsletter", "users"."email_unread_notifications", "users"."encrypted_password", "users"."experience_level", "users"."export_requested", "users"."exported_at", "users"."facebook_created_at", "users"."facebook_username", "users"."failed_attempts", "users"."feed_fetched_at", "users"."feed_mark_canonical", "users"."feed_referential_link", "users"."feed_url", "users"."following_orgs_count", "users"."following_tags_count", "users"."following_users_count", "users"."github_created_at", "users"."github_repos_updated_at", "users"."github_username", "users"."inbox_guidelines", "users"."inbox_type", "users"."invitation_accepted_at", "users"."invitation_created_at", "users"."invitation_limit", "users"."invitation_sent_at", "users"."invitation_token", "users"."invitations_count", "users"."invited_by_id", "users"."invited_by_type", "users"."last_article_at", "users"."last_comment_at", "users"."last_followed_at", "users"."last_moderation_notification", "users"."last_notification_activity", "users"."last_onboarding_page", "users"."last_reacted_at", "users"."last_sign_in_at", "users"."last_sign_in_ip", "users"."latest_article_updated_at", "users"."locked_at", "users"."mobile_comment_notifications", "users"."mod_roundrobin_notifications", "users"."monthly_dues", "users"."name", "users"."old_old_username", "users"."old_username", "users"."onboarding_package_requested", "users"."organization_info_updated_at", "users"."payment_pointer", "users"."permit_adjacent_sponsors", "users"."profile_image", "users"."profile_updated_at", "users"."rating_votes_count", "users"."reaction_notifications", "users"."reactions_count", "users"."registered", "users"."registered_at", "users"."remember_created_at", "users"."remember_token", "users"."reputation_modifier", "users"."reset_password_sent_at", "users"."reset_password_token", "users"."saw_onboarding", "users"."score", "users"."secret", "users"."sign_in_count", "users"."signup_cta_variant", "users"."spent_credits_count", "users"."stripe_id_code", "users"."subscribed_to_user_subscriptions_count", "users"."twitter_created_at", "users"."twitter_username", "users"."unconfirmed_email", "users"."unlock_token", "users"."unspent_credits_count", "users"."updated_at", "users"."username", "users"."welcome_notifications", "users"."workshop_expiration" FROM "users" WHERE "users"."id" = $1  [["id", 4001]]
  Organization Load (0.3ms)  SELECT "organizations".* FROM "organizations" WHERE "organizations"."id" = $1 LIMIT $2  [["id", 4], ["LIMIT", 1]]
  Follow Load (0.4ms)  SELECT "follows".* FROM "follows" WHERE "follows"."followable_id" = $1 AND "follows"."followable_type" = $2 AND "follows"."subscription_status" = $3  [["followable_id", 4], ["followable_type", "Organization"], ["subscription_status", "all_articles"]]
```

1. loads `user`
2. loads `followers_scoped` for the user
3. loads the `User` objects found through `.map()`
4. loads the organization
5. loads `followers_scoped` for the user
6. loads the `User` objects found through `.map()`

I'm proposing a slightly different approach which always requires 2 queries:

```suggestion
  def followers_union
    relation = Follow.followable_user(user_id)
    relation = relation.union(Follow.followable_organization(organization_id)) if organization_id
    relation = relation.where(subscription_status: "all_articles")
    User.where(id: relation.pluck(:follower_id))
  end
```

One query to load the `followers` from the table and another query to load the `users`. This always has the upside of returning a relation, which can be manipulated further if needed.

It requires an additional scope in the `Follow` model:

```ruby
  scope :followable_organization, ->(id) { where(followable_id: id, followable_type: "Organization") }
```

but you can replace it with the inline code if you don't need it.

This is the result:

```sql
[57] pry(main)> art.followers_union
   (0.8ms)  SELECT "follower_id" FROM ( (SELECT "follows".* FROM "follows" WHERE "follows"."followable_id" = $1 AND "follows"."followable_type" = $2) UNION (SELECT "follows".* FROM "follows" WHERE "follows"."followable_id" = $3 AND "follows"."followable_type" = $4) ) "follows" WHERE "follows"."subscription_status" = $5  [["followable_id", 4002], ["followable_type", "User"], ["followable_id", 4], ["followable_type", "Organization"], ["subscription_status", "all_articles"]]
  User Load (0.3ms)  SELECT "users"."id", "users"."apple_created_at", "users"."apple_username", "users"."articles_count", "users"."badge_achievements_count", "users"."blocked_by_count", "users"."blocking_others_count", "users"."checked_code_of_conduct", "users"."checked_terms_and_conditions", "users"."comments_count", "users"."config_font", "users"."config_navbar", "users"."config_theme", "users"."confirmation_sent_at", "users"."confirmation_token", "users"."confirmed_at", "users"."created_at", "users"."credits_count", "users"."current_sign_in_at", "users"."current_sign_in_ip", "users"."display_announcements", "users"."display_sponsors", "users"."editor_version", "users"."email", "users"."email_badge_notifications", "users"."email_comment_notifications", "users"."email_community_mod_newsletter", "users"."email_connect_messages", "users"."email_digest_periodic", "users"."email_follower_notifications", "users"."email_membership_newsletter", "users"."email_mention_notifications", "users"."email_newsletter", "users"."email_tag_mod_newsletter", "users"."email_unread_notifications", "users"."encrypted_password", "users"."experience_level", "users"."export_requested", "users"."exported_at", "users"."facebook_created_at", "users"."facebook_username", "users"."failed_attempts", "users"."feed_fetched_at", "users"."feed_mark_canonical", "users"."feed_referential_link", "users"."feed_url", "users"."following_orgs_count", "users"."following_tags_count", "users"."following_users_count", "users"."github_created_at", "users"."github_repos_updated_at", "users"."github_username", "users"."inbox_guidelines", "users"."inbox_type", "users"."invitation_accepted_at", "users"."invitation_created_at", "users"."invitation_limit", "users"."invitation_sent_at", "users"."invitation_token", "users"."invitations_count", "users"."invited_by_id", "users"."invited_by_type", "users"."last_article_at", "users"."last_comment_at", "users"."last_followed_at", "users"."last_moderation_notification", "users"."last_notification_activity", "users"."last_onboarding_page", "users"."last_reacted_at", "users"."last_sign_in_at", "users"."last_sign_in_ip", "users"."latest_article_updated_at", "users"."locked_at", "users"."mobile_comment_notifications", "users"."mod_roundrobin_notifications", "users"."monthly_dues", "users"."name", "users"."old_old_username", "users"."old_username", "users"."onboarding_package_requested", "users"."organization_info_updated_at", "users"."payment_pointer", "users"."permit_adjacent_sponsors", "users"."profile_image", "users"."profile_updated_at", "users"."rating_votes_count", "users"."reaction_notifications", "users"."reactions_count", "users"."registered", "users"."registered_at", "users"."remember_created_at", "users"."remember_token", "users"."reputation_modifier", "users"."reset_password_sent_at", "users"."reset_password_token", "users"."saw_onboarding", "users"."score", "users"."secret", "users"."sign_in_count", "users"."signup_cta_variant", "users"."spent_credits_count", "users"."stripe_id_code", "users"."subscribed_to_user_subscriptions_count", "users"."twitter_created_at", "users"."twitter_username", "users"."unconfirmed_email", "users"."unlock_token", "users"."unspent_credits_count", "users"."updated_at", "users"."username", "users"."welcome_notifications", "users"."workshop_expiration" FROM "users" WHERE "users"."id" = $1  [["id", 4001]]
```

Reply:
Checked in a prod console with both of these methods monkeypatched in:

```
irb(main):031:0> a.followers.to_a.count
D, [2021-04-06T16:16:22.078129 #3] DEBUG -- :   Follow Load (6.3ms)  SELECT "follows".* FROM "follows" WHERE "follows"."followable_id" = $1 AND "follows"."followable_type" = $2 AND "follows"."subscription_status" = $3  [["followable_id", 17908], ["followable_type", "User"], ["subscription_status", "all_articles"]]
D, [2021-04-06T16:16:22.114674 #3] DEBUG -- :   User Load (16.3ms)  SELECT "users"."id", "users"."name", "users"."twitter_username", "users"."github_username", "users"."created_at", "users"."updated_at", "users"."profile_image", "users"."email", "users"."encrypted_password", "users"."reset_password_token", "users"."reset_password_sent_at", "users"."remember_created_at", "users"."sign_in_count", "users"."current_sign_in_at", "users"."last_sign_in_at", "users"."current_sign_in_ip", "users"."last_sign_in_ip", "users"."confirmation_token", "users"."confirmed_at", "users"."confirmation_sent_at", "users"."unconfirmed_email", "users"."username", "users"."remember_token", "users"."secret", "users"."email_newsletter", "users"."email_comment_notifications", "users"."reputation_modifier", "users"."score", "users"."email_follower_notifications", "users"."signup_cta_variant", "users"."experience_level", "users"."onboarding_package_requested", "users"."github_created_at", "users"."twitter_created_at", "users"."feed_url", "users"."articles_count", "users"."comments_count", "users"."old_username", "users"."old_old_username", "users"."reactions_count", "users"."email_unread_notifications", "users"."feed_mark_canonical", "users"."last_moderation_notification", "users"."last_notification_activity", "users"."stripe_id_code", "users"."checked_code_of_conduct", "users"."email_mention_notifications", "users"."following_tags_count", "users"."following_users_count", "users"."following_orgs_count", "users"."saw_onboarding", "users"."monthly_dues", "users"."display_sponsors", "users"."email_membership_newsletter", "users"."workshop_expiration", "users"."permit_adjacent_sponsors", "users"."email_digest_periodic", "users"."last_followed_at", "users"."badge_achievements_count", "users"."email_badge_notifications", "users"."email_connect_messages", "users"."editor_version", "users"."export_requested", "users"."exported_at", "users"."mobile_comment_notifications", "users"."feed_fetched_at", "users"."last_article_at", "users"."last_comment_at", "users"."profile_updated_at", "users"."github_repos_updated_at", "users"."rating_votes_count", "users"."config_theme", "users"."config_font", "users"."inbox_type", "users"."email_tag_mod_newsletter", "users"."email_community_mod_newsletter", "users"."inbox_guidelines", "users"."credits_count", "users"."spent_credits_count", "users"."unspent_credits_count", "users"."feed_referential_link", "users"."checked_terms_and_conditions", "users"."last_onboarding_page", "users"."mod_roundrobin_notifications", "users"."blocked_by_count", "users"."blocking_others_count", "users"."config_navbar", "users"."organization_info_updated_at", "users"."welcome_notifications", "users"."display_announcements", "users"."payment_pointer", "users"."subscribed_to_user_subscriptions_count", "users"."invitation_token", "users"."registered", "users"."registered_at", "users"."invitation_created_at", "users"."invitation_sent_at", "users"."invitation_accepted_at", "users"."invitation_limit", "users"."invited_by_type", "users"."invited_by_id", "users"."invitations_count", "users"."failed_attempts", "users"."unlock_token", "users"."locked_at", "users"."reaction_notifications", "users"."facebook_username", "users"."facebook_created_at", "users"."last_reacted_at", "users"."apple_created_at", "users"."apple_username", "users"."latest_article_updated_at" FROM "users" WHERE "users"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67, $68, $69, $70, $71, $72, $73, $74, $75, $76, $77, $78, $79, $80, $81, $82, $83, $84, $85, $86, $87, $88, $89, $90, $91, $92, $93, $94, $95, $96, $97, $98, $99, $100, $101, $102, $103, $104, $105, $106, $107, $108, $109, $110, $111, $112, $113, $114, $115, $116, $117, $118, $119, $120, $121, $122, $123, $124, $125, $126, $127, $128, $129, $130, $131, $132, $133, $134, $135, $136, $137, $138, $139, $140, $141, $142, $143, $144, $145, $146, $147, $148, $149, $150, $151, $152, $153, $154, $155, $156, $157, $158, $159, $160, $161, $162, $163, $164, $165, $166, $167, $168, $169, $170, $171, $172, $173, $174, $175, $176, $177, $178, $179, $180, $181, $182, $183, $184, $185, $186, $187, $188, $189, $190, $191, $192, $193, $194, $195, $196, $197, $198, $199, $200, $201, $202, $203, $204, $205, $206, $207, $208, $209, $210, $211, $212, $213, $214, $215, $216, $217, $218, $219, $220, $221, $222, $223, $224, $225, $226, $227, $228, $229, $230, $231, $232, $233, $234, $235, $236, $237, $238, $239, $240, $241, $242, $243, $244, $245, $246, $247, $248, $249, $250, $251, $252, $253, $254, $255, $256, $257, $258, $259, $260, $261, $262, $263, $264, $265, $266, $267, $268, $269, $270, $271, $272, $273, $274, $275, $276, $277, $278, $279, $280, $281, $282, $283, $284, $285, $286, $287, $288, $289, $290, $291, $292, $293, $294, $295, $296, $297, $298, $299, $300, $301, $302, $303, $304, $305, $306, $307, $308, $309, $310, $311, $312, $313, $314, $315, $316, $317, $318, $319, $320, $321, $322, $323, $324, $325, $326, $327, $328, $329, $330, $331, $332, $333, $334, $335, $336, $337, $338, $339, $340, $341, $342, $343, $344, $345, $346, $347, $348, $349, $350, $351, $352, $353, $354, $355, $356, $357, $358, $359, $360, $361, $362, $363, $364, $365, $366, $367, $368, $369, $370, $371, $372, $373, $374, $375, $376, $377, $378, $379, $380, $381)  [[nil, 1], [nil, 13962], [nil, 510], [nil, 560], [nil, 1075], [nil, 2225], [nil, 241632], [nil, 78122], [nil, 78270], [nil, 84637], [nil, 241648], [nil, 241655], [nil, 241670], [nil, 98519], [nil, 98758], [nil, 98753], [nil, 98872], [nil, 99143], [nil, 99253], [nil, 99256], [nil, 99354], [nil, 99353], [nil, 99407], [nil, 99475], [nil, 99644], [nil, 99791], [nil, 100196], [nil, 100517], [nil, 100533], [nil, 102057], [nil, 105681], [nil, 241757], [nil, 241775], [nil, 241832], [nil, 241845], [nil, 241859], [nil, 129691], [nil, 132880], [nil, 132878], [nil, 111269], [nil, 134278], [nil, 134450], [nil, 134454], [nil, 136482], [nil, 136485], [nil, 138611], [nil, 140246], [nil, 140249], [nil, 241957], [nil, 241959], [nil, 136602], [nil, 241973], [nil, 145132], [nil, 241978], [nil, 148742], [nil, 148741], [nil, 45362], [nil, 242024], [nil, 151446], [nil, 242030], [nil, 238744], [nil, 155732], [nil, 157686], [nil, 164616], [nil, 16997], [nil, 167048], [nil, 167053], [nil, 167690], [nil, 171540], [nil, 173394], [nil, 173409], [nil, 174284], [nil, 176054], [nil, 176589], [nil, 177135], [nil, 177128], [nil, 177133], [nil, 242709], [nil, 180665], [nil, 180669], [nil, 242346], [nil, 243027], [nil, 242424], [nil, 242430], [nil, 242436], [nil, 242442], [nil, 242545], [nil, 243081], [nil, 242699], [nil, 243282], [nil, 243460], [nil, 243465], [nil, 246505], [nil, 246511], [nil, 246577], [nil, 246581], [nil, 196880], [nil, 197868], [nil, 250729], [nil, 200419], [nil, 251215], [nil, 250053], [nil, 250805], [nil, 250807], [nil, 250814], [nil, 250815], [nil, 249537], [nil, 572654], [nil, 255121], [nil, 206905], [nil, 207720], [nil, 207714], [nil, 207721], [nil, 207722], [nil, 258227], [nil, 258224], [nil, 258230], [nil, 210177], [nil, 210180], [nil, 259513], [nil, 259522], [nil, 213123], [nil, 213938], [nil, 265796], [nil, 219616], [nil, 267233], [nil, 267236], [nil, 268763], [nil, 268780], [nil, 222234], [nil, 222239], [nil, 273243], [nil, 273240], [nil, 273255], [nil, 273252], [nil, 225612], [nil, 225616], [nil, 232323], [nil, 232324], [nil, 232336], [nil, 232353], [nil, 232390], [nil, 232395], [nil, 232403], [nil, 232405], [nil, 232408], [nil, 234385], [nil, 234384], [nil, 234387], [nil, 234389], [nil, 234390], [nil, 234394], [nil, 234413], [nil, 234414], [nil, 234611], [nil, 234612], [nil, 234627], [nil, 234661], [nil, 234663], [nil, 235320], [nil, 235322], [nil, 235360], [nil, 235368], [nil, 235455], [nil, 235893], [nil, 235896], [nil, 235900], [nil, 235901], [nil, 71323], [nil, 235975], [nil, 236248], [nil, 236554], [nil, 236665], [nil, 236664], [nil, 215833], [nil, 237018], [nil, 237022], [nil, 237025], [nil, 237028], [nil, 237825], [nil, 284375], [nil, 284378], [nil, 239986], [nil, 239960], [nil, 240682], [nil, 335024], [nil, 336369], [nil, 336374], [nil, 387016], [nil, 300580], [nil, 300587], [nil, 300705], [nil, 300840], [nil, 300842], [nil, 300884], [nil, 300891], [nil, 301080], [nil, 301095], [nil, 301805], [nil, 303484], [nil, 303480], [nil, 303485], [nil, 303487], [nil, 303879], [nil, 346309], [nil, 347746], [nil, 347743], [nil, 311145], [nil, 63000], [nil, 311186], [nil, 311188], [nil, 237209], [nil, 311231], [nil, 311230], [nil, 311247], [nil, 158103], [nil, 311236], [nil, 311244], [nil, 200603], [nil, 119875], [nil, 311312], [nil, 134097], [nil, 311362], [nil, 28719], [nil, 40281], [nil, 311477], [nil, 173281], [nil, 298378], [nil, 311906], [nil, 311934], [nil, 346086], [nil, 346091], [nil, 311945], [nil, 311954], [nil, 311971], [nil, 312042], [nil, 70094], [nil, 312085], [nil, 312102], [nil, 312090], [nil, 312108], [nil, 312145], [nil, 255856], [nil, 348153], [nil, 312303], [nil, 312299], [nil, 312301], [nil, 312302], [nil, 312328], [nil, 312330], [nil, 312334], [nil, 312333], [nil, 312335], [nil, 312349], [nil, 312344], [nil, 312345], [nil, 196923], [nil, 312526], [nil, 312562], [nil, 312553], [nil, 312558], [nil, 312559], [nil, 237846], [nil, 124819], [nil, 313310], [nil, 354008], [nil, 119473], [nil, 314956], [nil, 354430], [nil, 354429], [nil, 354427], [nil, 354503], [nil, 316779], [nil, 316784], [nil, 354025], [nil, 9067], [nil, 77243], [nil, 318282], [nil, 354423], [nil, 361019], [nil, 319083], [nil, 319306], [nil, 319607], [nil, 319608], [nil, 173833], [nil, 322300], [nil, 125068], [nil, 91819], [nil, 325087], [nil, 325823], [nil, 89970], [nil, 155324], [nil, 425177], [nil, 419177], [nil, 420326], [nil, 105545], [nil, 88959], [nil, 148358], [nil, 425989], [nil, 173391], [nil, 418229], [nil, 424024], [nil, 260767], [nil, 118656], [nil, 41065], [nil, 419674], [nil, 419768], [nil, 424087], [nil, 422668], [nil, 596], [nil, 422757], [nil, 145020], [nil, 165494], [nil, 429375], [nil, 329514], [nil, 329510], [nil, 329511], [nil, 429395], [nil, 329538], [nil, 330701], [nil, 330703], [nil, 330705], [nil, 330870], [nil, 331601], [nil, 331605], [nil, 332247], [nil, 2693], [nil, 119279], [nil, 461206], [nil, 27836], [nil, 580007], [nil, 473648], [nil, 600594], [nil, 593489], [nil, 591767], [nil, 595327], [nil, 595341], [nil, 571015], [nil, 591933], [nil, 594015], [nil, 591945], [nil, 9597], [nil, 590504], [nil, 594081], [nil, 111154], [nil, 342525], [nil, 590700], [nil, 594348], [nil, 592217], [nil, 592241], [nil, 592347], [nil, 592467], [nil, 92143], [nil, 591165], [nil, 594093], [nil, 589280], [nil, 304290], [nil, 591396], [nil, 594969], [nil, 591451], [nil, 57772], [nil, 542594], [nil, 593350], [nil, 595185], [nil, 603364], [nil, 601326], [nil, 601333], [nil, 603574], [nil, 597753], [nil, 598105], [nil, 601823], [nil, 598300], [nil, 164389], [nil, 600030], [nil, 602727], [nil, 602817], [nil, 603052], [nil, 603058], [nil, 603211], [nil, 601096], [nil, 500845]]
=> 381
irb(main):032:0> a.followers_union.to_a.count
D, [2021-04-06T16:17:10.699123 #3] DEBUG -- :    (2.7ms)  SELECT "follows"."follower_id" FROM "follows" WHERE "follows"."followable_id" = $1 AND "follows"."followable_type" = $2 AND "follows"."subscription_status" = $3  [["followable_id", 17908], ["followable_type", "User"], ["subscription_status", "all_articles"]]
D, [2021-04-06T16:17:10.713008 #3] DEBUG -- :   User Load (10.0ms)  SELECT "users"."id", "users"."name", "users"."twitter_username", "users"."github_username", "users"."created_at", "users"."updated_at", "users"."profile_image", "users"."email", "users"."encrypted_password", "users"."reset_password_token", "users"."reset_password_sent_at", "users"."remember_created_at", "users"."sign_in_count", "users"."current_sign_in_at", "users"."last_sign_in_at", "users"."current_sign_in_ip", "users"."last_sign_in_ip", "users"."confirmation_token", "users"."confirmed_at", "users"."confirmation_sent_at", "users"."unconfirmed_email", "users"."username", "users"."remember_token", "users"."secret", "users"."email_newsletter", "users"."email_comment_notifications", "users"."reputation_modifier", "users"."score", "users"."email_follower_notifications", "users"."signup_cta_variant", "users"."experience_level", "users"."onboarding_package_requested", "users"."github_created_at", "users"."twitter_created_at", "users"."feed_url", "users"."articles_count", "users"."comments_count", "users"."old_username", "users"."old_old_username", "users"."reactions_count", "users"."email_unread_notifications", "users"."feed_mark_canonical", "users"."last_moderation_notification", "users"."last_notification_activity", "users"."stripe_id_code", "users"."checked_code_of_conduct", "users"."email_mention_notifications", "users"."following_tags_count", "users"."following_users_count", "users"."following_orgs_count", "users"."saw_onboarding", "users"."monthly_dues", "users"."display_sponsors", "users"."email_membership_newsletter", "users"."workshop_expiration", "users"."permit_adjacent_sponsors", "users"."email_digest_periodic", "users"."last_followed_at", "users"."badge_achievements_count", "users"."email_badge_notifications", "users"."email_connect_messages", "users"."editor_version", "users"."export_requested", "users"."exported_at", "users"."mobile_comment_notifications", "users"."feed_fetched_at", "users"."last_article_at", "users"."last_comment_at", "users"."profile_updated_at", "users"."github_repos_updated_at", "users"."rating_votes_count", "users"."config_theme", "users"."config_font", "users"."inbox_type", "users"."email_tag_mod_newsletter", "users"."email_community_mod_newsletter", "users"."inbox_guidelines", "users"."credits_count", "users"."spent_credits_count", "users"."unspent_credits_count", "users"."feed_referential_link", "users"."checked_terms_and_conditions", "users"."last_onboarding_page", "users"."mod_roundrobin_notifications", "users"."blocked_by_count", "users"."blocking_others_count", "users"."config_navbar", "users"."organization_info_updated_at", "users"."welcome_notifications", "users"."display_announcements", "users"."payment_pointer", "users"."subscribed_to_user_subscriptions_count", "users"."invitation_token", "users"."registered", "users"."registered_at", "users"."invitation_created_at", "users"."invitation_sent_at", "users"."invitation_accepted_at", "users"."invitation_limit", "users"."invited_by_type", "users"."invited_by_id", "users"."invitations_count", "users"."failed_attempts", "users"."unlock_token", "users"."locked_at", "users"."reaction_notifications", "users"."facebook_username", "users"."facebook_created_at", "users"."last_reacted_at", "users"."apple_created_at", "users"."apple_username", "users"."latest_article_updated_at" FROM "users" WHERE "users"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67, $68, $69, $70, $71, $72, $73, $74, $75, $76, $77, $78, $79, $80, $81, $82, $83, $84, $85, $86, $87, $88, $89, $90, $91, $92, $93, $94, $95, $96, $97, $98, $99, $100, $101, $102, $103, $104, $105, $106, $107, $108, $109, $110, $111, $112, $113, $114, $115, $116, $117, $118, $119, $120, $121, $122, $123, $124, $125, $126, $127, $128, $129, $130, $131, $132, $133, $134, $135, $136, $137, $138, $139, $140, $141, $142, $143, $144, $145, $146, $147, $148, $149, $150, $151, $152, $153, $154, $155, $156, $157, $158, $159, $160, $161, $162, $163, $164, $165, $166, $167, $168, $169, $170, $171, $172, $173, $174, $175, $176, $177, $178, $179, $180, $181, $182, $183, $184, $185, $186, $187, $188, $189, $190, $191, $192, $193, $194, $195, $196, $197, $198, $199, $200, $201, $202, $203, $204, $205, $206, $207, $208, $209, $210, $211, $212, $213, $214, $215, $216, $217, $218, $219, $220, $221, $222, $223, $224, $225, $226, $227, $228, $229, $230, $231, $232, $233, $234, $235, $236, $237, $238, $239, $240, $241, $242, $243, $244, $245, $246, $247, $248, $249, $250, $251, $252, $253, $254, $255, $256, $257, $258, $259, $260, $261, $262, $263, $264, $265, $266, $267, $268, $269, $270, $271, $272, $273, $274, $275, $276, $277, $278, $279, $280, $281, $282, $283, $284, $285, $286, $287, $288, $289, $290, $291, $292, $293, $294, $295, $296, $297, $298, $299, $300, $301, $302, $303, $304, $305, $306, $307, $308, $309, $310, $311, $312, $313, $314, $315, $316, $317, $318, $319, $320, $321, $322, $323, $324, $325, $326, $327, $328, $329, $330, $331, $332, $333, $334, $335, $336, $337, $338, $339, $340, $341, $342, $343, $344, $345, $346, $347, $348, $349, $350, $351, $352, $353, $354, $355, $356, $357, $358, $359, $360, $361, $362, $363, $364, $365, $366, $367, $368, $369, $370, $371, $372, $373, $374, $375, $376, $377, $378, $379, $380, $381)  [[nil, 1], [nil, 13962], [nil, 510], [nil, 560], [nil, 1075], [nil, 2225], [nil, 241632], [nil, 78122], [nil, 78270], [nil, 84637], [nil, 241648], [nil, 241655], [nil, 241670], [nil, 98519], [nil, 98758], [nil, 98753], [nil, 98872], [nil, 99143], [nil, 99253], [nil, 99256], [nil, 99354], [nil, 99353], [nil, 99407], [nil, 99475], [nil, 99644], [nil, 99791], [nil, 100196], [nil, 100517], [nil, 100533], [nil, 102057], [nil, 105681], [nil, 241757], [nil, 241775], [nil, 241832], [nil, 241845], [nil, 241859], [nil, 129691], [nil, 132880], [nil, 132878], [nil, 111269], [nil, 134278], [nil, 134450], [nil, 134454], [nil, 136482], [nil, 136485], [nil, 138611], [nil, 140246], [nil, 140249], [nil, 241957], [nil, 241959], [nil, 136602], [nil, 241973], [nil, 145132], [nil, 241978], [nil, 148742], [nil, 148741], [nil, 45362], [nil, 242024], [nil, 151446], [nil, 242030], [nil, 238744], [nil, 155732], [nil, 157686], [nil, 164616], [nil, 16997], [nil, 167048], [nil, 167053], [nil, 167690], [nil, 171540], [nil, 173394], [nil, 173409], [nil, 174284], [nil, 176054], [nil, 176589], [nil, 177135], [nil, 177128], [nil, 177133], [nil, 242709], [nil, 180665], [nil, 180669], [nil, 242346], [nil, 243027], [nil, 242424], [nil, 242430], [nil, 242436], [nil, 242442], [nil, 242545], [nil, 243081], [nil, 242699], [nil, 243282], [nil, 243460], [nil, 243465], [nil, 246505], [nil, 246511], [nil, 246577], [nil, 246581], [nil, 196880], [nil, 197868], [nil, 250729], [nil, 200419], [nil, 251215], [nil, 250053], [nil, 250805], [nil, 250807], [nil, 250814], [nil, 250815], [nil, 249537], [nil, 572654], [nil, 255121], [nil, 206905], [nil, 207720], [nil, 207714], [nil, 207721], [nil, 207722], [nil, 258227], [nil, 258224], [nil, 258230], [nil, 210177], [nil, 210180], [nil, 259513], [nil, 259522], [nil, 213123], [nil, 213938], [nil, 265796], [nil, 219616], [nil, 267233], [nil, 267236], [nil, 268763], [nil, 268780], [nil, 222234], [nil, 222239], [nil, 273243], [nil, 273240], [nil, 273255], [nil, 273252], [nil, 225612], [nil, 225616], [nil, 232323], [nil, 232324], [nil, 232336], [nil, 232353], [nil, 232390], [nil, 232395], [nil, 232403], [nil, 232405], [nil, 232408], [nil, 234385], [nil, 234384], [nil, 234387], [nil, 234389], [nil, 234390], [nil, 234394], [nil, 234413], [nil, 234414], [nil, 234611], [nil, 234612], [nil, 234627], [nil, 234661], [nil, 234663], [nil, 235320], [nil, 235322], [nil, 235360], [nil, 235368], [nil, 235455], [nil, 235893], [nil, 235896], [nil, 235900], [nil, 235901], [nil, 71323], [nil, 235975], [nil, 236248], [nil, 236554], [nil, 236665], [nil, 236664], [nil, 215833], [nil, 237018], [nil, 237022], [nil, 237025], [nil, 237028], [nil, 237825], [nil, 284375], [nil, 284378], [nil, 239986], [nil, 239960], [nil, 240682], [nil, 335024], [nil, 336369], [nil, 336374], [nil, 387016], [nil, 300580], [nil, 300587], [nil, 300705], [nil, 300840], [nil, 300842], [nil, 300884], [nil, 300891], [nil, 301080], [nil, 301095], [nil, 301805], [nil, 303484], [nil, 303480], [nil, 303485], [nil, 303487], [nil, 303879], [nil, 346309], [nil, 347746], [nil, 347743], [nil, 311145], [nil, 63000], [nil, 311186], [nil, 311188], [nil, 237209], [nil, 311231], [nil, 311230], [nil, 311247], [nil, 158103], [nil, 311236], [nil, 311244], [nil, 200603], [nil, 119875], [nil, 311312], [nil, 134097], [nil, 311362], [nil, 28719], [nil, 40281], [nil, 311477], [nil, 173281], [nil, 298378], [nil, 311906], [nil, 311934], [nil, 346086], [nil, 346091], [nil, 311945], [nil, 311954], [nil, 311971], [nil, 312042], [nil, 70094], [nil, 312085], [nil, 312102], [nil, 312090], [nil, 312108], [nil, 312145], [nil, 255856], [nil, 348153], [nil, 312303], [nil, 312299], [nil, 312301], [nil, 312302], [nil, 312328], [nil, 312330], [nil, 312334], [nil, 312333], [nil, 312335], [nil, 312349], [nil, 312344], [nil, 312345], [nil, 196923], [nil, 312526], [nil, 312562], [nil, 312553], [nil, 312558], [nil, 312559], [nil, 237846], [nil, 124819], [nil, 313310], [nil, 354008], [nil, 119473], [nil, 314956], [nil, 354430], [nil, 354429], [nil, 354427], [nil, 354503], [nil, 316779], [nil, 316784], [nil, 354025], [nil, 9067], [nil, 77243], [nil, 318282], [nil, 354423], [nil, 361019], [nil, 319083], [nil, 319306], [nil, 319607], [nil, 319608], [nil, 173833], [nil, 322300], [nil, 125068], [nil, 91819], [nil, 325087], [nil, 325823], [nil, 89970], [nil, 155324], [nil, 425177], [nil, 419177], [nil, 420326], [nil, 105545], [nil, 88959], [nil, 148358], [nil, 425989], [nil, 173391], [nil, 418229], [nil, 424024], [nil, 260767], [nil, 118656], [nil, 41065], [nil, 419674], [nil, 419768], [nil, 424087], [nil, 422668], [nil, 596], [nil, 422757], [nil, 145020], [nil, 165494], [nil, 429375], [nil, 329514], [nil, 329510], [nil, 329511], [nil, 429395], [nil, 329538], [nil, 330701], [nil, 330703], [nil, 330705], [nil, 330870], [nil, 331601], [nil, 331605], [nil, 332247], [nil, 2693], [nil, 119279], [nil, 461206], [nil, 27836], [nil, 580007], [nil, 473648], [nil, 600594], [nil, 593489], [nil, 591767], [nil, 595327], [nil, 595341], [nil, 571015], [nil, 591933], [nil, 594015], [nil, 591945], [nil, 9597], [nil, 590504], [nil, 594081], [nil, 111154], [nil, 342525], [nil, 590700], [nil, 594348], [nil, 592217], [nil, 592241], [nil, 592347], [nil, 592467], [nil, 92143], [nil, 591165], [nil, 594093], [nil, 589280], [nil, 304290], [nil, 591396], [nil, 594969], [nil, 591451], [nil, 57772], [nil, 542594], [nil, 593350], [nil, 595185], [nil, 603364], [nil, 601326], [nil, 601333], [nil, 603574], [nil, 597753], [nil, 598105], [nil, 601823], [nil, 598300], [nil, 164389], [nil, 600030], [nil, 602727], [nil, 602817], [nil, 603052], [nil, 603058], [nil, 603211], [nil, 601096], [nil, 500845]]
=> 381
```

@rhymes I _really_ like the idea of returning an ActiveRecord relation instead of an array, but your solution doesn't seem to involve fewer queries unless it's under an organization. Vaidehi's query _only_ loads the org if the article was posted under one, so [for 96% of articles](https://dev.to/admin/blazer/queries/406-org-vs-non-org-articles-in-the-past-year) query times for followers should be nearly identical between the two.

It does make a lot of sense to return a relation. However, it's likely beyond the scope of this PR, the purpose of which is to fix a [Feature Envy code smell](https://wiki.c2.com/?FeatureEnvySmell).

Reply:
@jgaskins I'm a bit confused about a detail: I counted 3 queries for articles without org and 5-6 with the org. The union based implementation still skips loading the org followers if they aren't needed. And it amounts to 2 query in each case.

If we want to return an array, we can still call `.to_a` on the last line of the method

Let me know as I don't really see the drawback in my proposal :D

Reply:
Could this be a separate `internalEngineering` issue that we could prioritize? Maybe it's even a SRE task hehe :trollface: 

I would love to keep this extraction separate and do any refactoring as a separate PR -- I say this purely because I don't want to be too derailed from RFC 22 (which this PR is in service of!) by refactoring this method and doing the due diligence of testing it out and watching it after it goes out to production/the fleet 😅

Reply:
@rhymes 

> I counted 3 queries for articles without org

It looks like 3 when you're querying it in a vacuum, but the first is the query for the article's author, which will already be loaded.

This is only invoked by the `Articles::Creator`, which [assigns the association for the already loaded user](https://github.com/forem/forem/blob/b2ded5b889bf75b1023a93e2c6d20ace60713210/app/services/articles/creator.rb#L60), and `Articles::Updater`, which [calls `article.user`](https://github.com/forem/forem/blob/b2ded5b889bf75b1023a93e2c6d20ace60713210/app/services/articles/updater.rb#L22), loading the user before invoking `Notification.send_to_followers`.

So, realistically, it's only 2 queries. 🙂

> 5-6 with the org

Less than 4% of articles are posted within an org. It's rare enough (and those queries are fast enough) that a couple extra queries is fine as long as they're not prohibitively slow. From what I can tell, these are plenty fast.

> The union based implementation still skips loading the org followers if they aren't needed. … Let me know as I don't really see the drawback in my proposal :D

Apologies, I didn't mean to imply that your solution is bad or in any way worse than the existing implementation — the truth is that I don't know, so I can't make that judgement. What I meant was that it's fine to leave it since she's just moving code and it's fast enough.

Modifying code and _especially_ optimizing has costs (I gave [a RubyConf talk about this](https://www.youtube.com/watch?v=oze1Hu2sUdI) a few years ago), so we should be clear about what the benefits and costs are. In this case, the benefit is unclear and the cost is changing code that is highly stable (it hasn't changed for a reason other than Rubocop in 2 years) and that makes me nervous. :joy:

To clarify the "unclear benefit" part, the number of queries can be a deceptive metric — I get tripped up by it, too. I've seen plenty of occasions where N+1 queries are faster than eager loading and others where `JOIN`s are slower than multiple queries. And let me tell you that is _confusing af_. I am ecstatic that you're paying attention to performance (seriously, I've worked with a lot of experienced engineers that don't) but it's more complex than the number of queries. 🙂

When I look at [traces of the Sidekiq worker that executes this code](https://app.datadoghq.com/apm/resource/sidekiq/sidekiq.job/9e463b2a70b1e44a?end=1617741456234&env=production&index=apm-search&paused=true&query=env%3Aproduction%20service%3Asidekiq%20operation_name%3Asidekiq.job%20resource_name%3A%22Notifications%3A%3ANotifiableActionWorker%22%20%40duration%3A%3E%3D3s&spanID=3095106777556259622&start=1617136656234), the largest SQL segment of the flame graphs are almost always inserting into `notifications` or selecting user records by primary key[1].

<img width="506" alt="Screenshot of a flame graph of a Notifications::NotifiableActionWorker. The largest SQL query is inserting a notification row." src="https://user-images.githubusercontent.com/108205/113775412-a9cf7a00-96f6-11eb-92b1-fee177af65a1.png">

If we want this to be faster, we could probably figure out what's going on in that empty region between the instantiation of the 4,204 `User` models and the insert. It's probably just Ruby code inside the worker, and Vaidehi did optimize some of it by [removing one of the intermediate arrays](https://github.com/forem/forem/pull/13229/files/5f56a10239f653276f4dea16dac41fe14a2d027f#diff-31de3da49240e43896343b24c13009757946fd0e71299118144ed89554255c16R28), so maybe that will help.

---

[1] The fact that we often spend too much time in simple queries like selecting users by a group of ids here is part of a larger issue we're seeing with Postgres where our query-time distribution has super long tails — distance between p95->p99 is 2-3 orders of magnitude larger than p50->p95 distance across a lot of trivial queries. I've been investigating why that is and I kiiiiinda wanna blame Heroku but it may also be us. Not super important here but wanted to at least clarify.

Reply:
> To clarify the "unclear benefit" part, the number of queries can be a deceptive metric — I get tripped up by it, too. I've seen plenty of occasions where N+1 queries are faster than eager loading and others where JOINs are slower than multiple queries. And let me tell you that is confusing af. I am ecstatic that you're paying attention to performance (seriously, I've worked with a lot of experienced engineers that don't) but it's more complex than the number of queries. 🙂

Definitely! Thank you for reminding me that we shouldn't optimize without measuring first. I sometimes forget about that very important part of writing performance oriented code. 

Also thank you a lot for your ever thorough explanations and data-oriented approach!

Reply:
@rhymes And I appreciate you pushing me to communicate more clearly 🎉  💯